swarm/swap/protocol.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/log"
+)
+
+const (
+	IsActiveProtocol = true
+)
+
+type SwapProtocol struct {
+	peersMu sync.RWMutex
+	peers   map[discover.NodeID]*SwapProtocolPeer
+	swap    *Swap
+}
+
+// Peer is the Peer extension for the streaming protocol
+type SwapProtocolPeer struct {
+	*protocols.Peer
+	swapProtocol *SwapProtocol
+}
+
+func NewSwapProtocol(swapAccount *Swap) *SwapProtocol {
+	proto := &SwapProtocol{
+		peers: make(map[discover.NodeID]*SwapProtocolPeer),
+		swap:  swapAccount,
+	}
+	return proto
+}
+
+// NewPeer is the constructor for Peer
+func NewPeer(peer *protocols.Peer, swap *SwapProtocol) *SwapProtocolPeer {
+	p := &SwapProtocolPeer{
+		Peer:         peer,
+		swapProtocol: swap,
+	}
+	return p
+}
+
+type IssueChequeMsg struct {
+	Cheque *Cheque
+}
+
+type RedeemChequeMsg struct {
+}
+
+/////////////////////////////////////////////////////////////////////
+// SECTION: node.Service interface
+/////////////////////////////////////////////////////////////////////
```

Comment:
the Service should be implemented by the `Swap` type no? calling this service Protocol is misleading and not how we use it right?

Reply:
This is related to the question above. 

I am not sure I understand how this should be done in your view. This service interface is currently only for  the sending/receiving cheques protocol.